File: consumerproducer/src/main/java/com/example/consumerproducer/MainActivity.java
package com.example.consumerproducer;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        SyncStack stack = new SyncStack();

        Producer p1 = new Producer(stack);
        Thread proT1 = new Thread(p1);
        proT1.start();

        Producer p2 = new Producer(stack);
        Thread proT2 = new Thread(p2);
        proT2.start();

        Consumer c1 = new Consumer(stack);
        Thread consT1 = new Thread(c1);
        consT1.start();

        Consumer c2 = new Consumer(stack);
        Thread consT2 = new Thread(c2);
        consT2.start();
    }

}

File: study_abroad/coroutine/src/main/java/com/study/coroutine/MainActivity.kt
package com.study.coroutine

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import com.study.coroutine.lua.Coroutine
import com.study.coroutine.lua.Dispatcher
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        GlobalScope.launch {
            testLuaImitate()
        }

    }

    private suspend fun testLuaImitate(){
        val producer = Coroutine.create<Unit,Int>(Dispatcher()){
            for(i in 0..3){
                Log.e("TAG","${Thread.currentThread().name} send $i")
                yield(i)
            }
            200
        }
        val consumer = Coroutine.create<Int,Unit>(Dispatcher()) {param:Int ->
            Log.e("TAG","${Thread.currentThread().name} start $param")
            for(i in 0..3){
                val value = yield(Unit)
                Log.e("TAG","${Thread.currentThread().name} receive $value")
            }
        }
        //循环搭的引擎
        while(producer.isActive && consumer.isActive){
            val result = producer.resume(Unit)
            consumer.resume(result)
        }
    }

}